File: 程序员面试金典第6版/面试题 05.07. Exchange LCCI.py
# 0xAAAAAAAA (Ob1010101010101010) 相与 取得偶数位
# 0x55555555 (0b0101010101010101) 相与 取得奇数位
# 偶数部分右移一位 奇数部分左移一位
# 然后取或即可
# class Solution:
#     def exchangeBits(self, num: int) -> int:
#         odd = num & 0x55555555
#         odd <<= 1
#         even = num & 0xAAAAAAAA
#         even >>= 1

#         return odd | even

class Solution:
    def exchangeBits(self, num: int) -> int:
        return ((num & 0x55555555) << 1) | ((num & 0xAAAAAAAA) >> 1)

File: 程序员面试金典第6版/面试题 05.07. Exchange LCCI.java
class Solution {
    public int exchangeBits(int num) {
        int[] arr = new int[32];
        int bit = 0;
        while(num != 0) {
            arr[bit++] = num % 2;      // 高位放到数组前面了 低位放到数组后面了
            num /= 2;
        }
        for(int i=1; i<=bit; i += 2) { // 每次移2位 注意
            int t = arr[i];
            arr[i] = arr[i-1];
            arr[i-1] = t;
        }
        int res = 0;
        for(int i=bit; i>=0; i--) {    // 从后向前 先处理低位 从二进制数组恢复到整数
            res = res*2 + arr[i]; // res << 1 + arr[i]
        }
        return res;
    }
}